external/vcm/vcm/_preloaded.py

```diff
@@ -0,0 +1,61 @@
+from typing import Sequence, Callable, Any
+import collections.abc
+import random
+import concurrent.futures
+
+
+class Preloaded(collections.abc.Iterator):
+    """
+    Iterator for data which asynchronously pre-loads the next set of output.
+    """
+
+    def __init__(
+        self,
+        filenames: Sequence[str],
+        loader_function: Callable[[str], Any],
+        shuffle: bool = True,
+    ):
+        """
+        Args:
+            filenames: sequence to be passed to loader function
+            shuffle: if True, return items in random order
+        """
+        self.loader_function = loader_function
+        self._filenames = list(filenames)
+        self._shuffle = shuffle
+        self._maybe_shuffle()
```

Comment:
Ah this is good to point out. It may or may not be possible - the issue is the shuffling must happen each time a new iterator is created (each for loop needs a different random order).

The code isn't written as a `Map` at the moment, so doing what you wrote exactly would require completely re-writing this feature. If I switch the order, I can do `ThreadedMap(func, iterator_function)` where `def iterator_function: return shuffled(args)`.

Reply:
We agree this change is optional.